File: backend/app/models/publication.py
from datetime import datetime

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import relationship

from app.db.database import Base


class Publication(Base):
    __tablename__ = "publications"
    __table_args__ = (
        UniqueConstraint("clip_id", "platform", name="uq_publication_clip_platform"),
    )

    id = Column(Integer, primary_key=True, index=True)
    clip_id = Column(Integer, ForeignKey("clips.id"), nullable=False)
    publication_account_id = Column(Integer, ForeignKey("publication_accounts.id"), nullable=True)
    platform = Column(String, nullable=False, default="YOUTUBE")
    status = Column(String, nullable=False, default="PENDING")
    error_type = Column(String, nullable=True)
    error_details = Column(Text, nullable=True)
    attempts = Column(Integer, nullable=False, default=0)
    next_retry = Column(DateTime, nullable=True)
    scheduled_at = Column(DateTime, nullable=True)
    manual = Column(Boolean, nullable=False, default=False)
    platform_post_id = Column(String, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    published_at = Column(DateTime, nullable=True)
    thumbnail_uploaded_at = Column(DateTime, nullable=True)

    clip = relationship("Clip", back_populates="publications")
    publication_account = relationship("PublicationAccount", back_populates="publications")


class PublicationAccount(Base):
    __tablename__ = "publication_accounts"

    id = Column(Integer, primary_key=True, index=True)
    platform = Column(String, nullable=False, default="YOUTUBE")
    account_name = Column(String, nullable=False)
    platform_account_id = Column(String, nullable=False)
    enabled = Column(Boolean, nullable=False, default=True)
    is_default = Column(Boolean, nullable=False, default=False)
    credential_path = Column(String, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    publications = relationship("Publication", back_populates="publication_account")

File: frontend/src/types/publication.ts
export type PublicationStatus = "SCHEDULED" | "PENDING" | "UPLOADING" | "PROCESSING" | "WAITING_RETRY" | "PUBLISHED" | "FAILED" | "CANCELLED";

export interface PublicationSummary {
  status: PublicationStatus;
  total: number;
}

export interface PublicationAccount {
  id: number;
  platform: string;
  account_name: string;
  platform_account_id: string;
  enabled: boolean;
  is_default: boolean;
}

export interface Publication {
  id: number;
  clip_id: number;
  title?: string | null;
  duration?: number | null;
  thumbnail_url?: string | null;
  platform: string;
  account?: PublicationAccount | null;
  status: PublicationStatus;
  error_type?: string | null;
  error_details?: string | null;
  created_at?: string;
  updated_at?: string;
  published_at?: string | null;
  scheduled_at?: string | null;
  next_retry?: string | null;
  manual?: boolean;
  platform_post_id?: string | null;
  publication_url?: string | null;
  timezone?: string | null;
  attempts?: number;
}

export interface PublicationSettings {
  youtube_auto_publish: boolean;
  max_uploads_per_day: number;
  publish_schedule: string[];
  publish_timezone: string;
  manual_upload_counts_toward_daily_limit: boolean;
  tiktok_enabled: boolean;
}

export interface PublicationScheduleDay {
  date: string;
  count: number;
  times: string[];
  clip_ids: number[];
}

export interface PublicationSchedulePlan {
  total_clips: number;
  max_per_day: number;
  start_date: string;
  times: string[];
  estimated_days: number;
  days: PublicationScheduleDay[];
  scheduled: { clip_id: number; date: string; time: string; scheduled_at: string }[];
  created_publications?: number;
  publication_ids?: number[];
}

export interface PublicationSchedulePayload {
  clip_ids: number[];
  platform?: string;
  max_per_day: number;
  start_date: string;
  times: string[];
}

export interface YouTubeAccount {
  connected: boolean;
  status: "CONNECTED" | "DISCONNECTED" | "INVALID" | "CONNECTING" | "ERROR";
  channel_name?: string;
  channel_title?: string;
  channel_thumbnail?: string | null;
  channel_id?: string;
  account?: PublicationAccount;
}
